### PR TITLE
Poulet4 ccomp v1 runtime

### PR DESCRIPTION
**Pull request by nikaido-shinku**
Sorry this pr is longer than it should have been. My main goal is to build something that can automatically run test cases. And there are three subgoals here. I started working a little bit on all of them.

first: I need to implement extract. And for that purpose, I needed to record which variable is of what extract type. I also needed to change the structure of the code to enable implementing extern-related functions in a separate coq file. [Related Files: files under poulet4/Ccomp/ and poulet_Ccomp ]

second: I want to write a pretty printer. I figured it would be better to write it in Ocaml rather than in coq. [Related Files: all the ocaml files]

third: there's a mismatch between the representation of a p4cub program that the Ccomp expects and the p4cub program that ToP4cub generate. ToP4cub creates a declaration context, while I was only expecting a single TopDecl.d. I currently can't figure out a way to translate the former into the latter, because I don't know which TopDecl.d should I plug all the Control.d into. We really should discuss this.

### CODE REVIEW

File: bin/main.ml [hunk 1]
```diff
@@ -71,9 +71,10 @@ let check_command =
      +> flag "-typed-json" no_arg ~doc:" Pretty-print typed AST JSON"
      +> flag "-gen-loc" no_arg ~doc:" Generate locators in AST"
      +> flag "-printp4" no_arg ~doc:" Print checked syntax in P4"
+     +> flag "-printp4cup" no_arg ~doc: "Print the p4cub AST"
      +> flag "-printp4-file" (optional_with_default "out.p4" string) ~doc:"Path to print checked syntax in P4"
      +> anon ("p4file" %: string))
-    (fun verbose include_dir json pretty exportp4 normalize export_file typed_json gen_loc printp4 printp4_file p4file () ->
+    (fun verbose include_dir json pretty exportp4 normalize export_file typed_json gen_loc printp4 printp4cup printp4_file p4file () ->
```
Comment (hackedy) [inline bin/main.ml:77]: typo, should be printp4cub?

Comment (nikaido-shinku) [reply bin/main.ml:77]: oh right, I should rename those into printp4cub

File: Makefile [hunk 2]
```diff
@@ -20,9 +20,9 @@ WEB_EXAMPLES+=stf-test/custom-stf-tests/register.p4
 
 .PHONY: all deps build claims clean test test-stf web
 
-all: build
+all: deps build
 
-build: deps
+build:
```
Comment (hackedy) [inline Makefile:25]: build won't work without deps as a dependency I think.

File: deps/poulet4_Ccomp/Petr4Runtime.h [hunk 3]
```diff
@@ -7,12 +7,13 @@
...
 enum p4int {FIXBIT, FIXINT};
 
-struct packet_in {
-  void *in;
-};
-struct packet_out{
-  void *out;
-};
+typedef struct packet_in {
+  int *in;
```
Comment (hackedy) [inline deps/poulet4_Ccomp/Petr4Runtime.h:11]: should this be unsigned char and not int?

File: deps/poulet4_Ccomp/Petr4Runtime.h [hunk 4]
```diff
@@ -101,6 +101,25 @@ void init_bitvec(struct BitVec *dst, int sign, int w, char *val){
...
+    *data = true;
+  } else {
+    *data = false;
+  }
+  pkt->in ++;
+}
+void extract_bitvec(packet_in *pkt, BitVec *data, int is_signed, int width){
+  char val[data->width]; 
+  for(int i = 0; i < data->width; i++){
+    val[i] = (*(pkt->in)) + 48;
```
Comment (hackedy) [inline deps/poulet4_Ccomp/Petr4Runtime.h:116]: What is this doing?

**Review by hackedy — CHANGES_REQUESTED** [review]
bin/main.ml is not building right now. If you can fix that go ahead and merge, my comments are just nits and questions.

Here's the build error.
```
File "bin/main.ml", line 79, characters 76-87:
79 |        ignore (check_file include_dir p4file json pretty exportp4 normalize export_file typed_json gen_loc verbose printp4 printp4_file))
                                                                                 ^^^^^^^^^^^
Error: This expression has type string but an expression was expected of type
         bool
```

File: deps/poulet4_Ccomp/Petr4Runtime.h (hunk 3 above)
Comment (nikaido-shinku) [reply deps/poulet4_Ccomp/Petr4Runtime.h:11]: yes you are right

**Pull request merged**